xdsl/dialects/llvm.py

```diff
@@ -46,6 +48,123 @@ def parse_parameters(parser: BaseParser) -> list[Attribute]:
         return [StringAttr.from_str(""), ArrayAttr.from_list(params)]
 
 
+@irdl_attr_definition
+class LLVMPointerType(ParametrizedAttribute, MLIRType):
+    name = "llvm.ptr"
+
+    type: ParameterDef[Attribute | NoneAttr]
+    addr_space: ParameterDef[IntAttr | NoneAttr]
+
+    def print_parameters(self, printer: Printer) -> None:
+        if isinstance(self.type, NoneAttr):
+            return
+
+        printer.print_string("<")
+        printer.print_attribute(self.type)
+        if not isinstance(self.addr_space, NoneAttr):
+            printer.print_string(", ")
+            printer.print_attribute(self.addr_space)
+
+        printer.print_string(">")
+
+    @classmethod
+    def untyped(cls):
+        return cls([NoneAttr(), NoneAttr()])
+
+    @classmethod
+    def typed(cls, type: Attribute):
+        return cls([type, NoneAttr()])
+
+
+@irdl_op_definition
+class AllocaOp(Operation):
+    name = "llvm.alloca"
+
+    size: Annotated[Operand, IntegerType]
+
+    alignment: OpAttr[AnyIntegerAttr]
+
+    res: Annotated[OpResult, LLVMPointerType]
+
+    @classmethod
+    def get(cls,
+            size: SSAValue | Operation,
+            elem_type: Attribute,
+            alignment: int = 32,
+            as_untyped_ptr: bool = False):
+        attrs: dict[str, Attribute] = {
+            'alignment': IntegerAttr.from_int_and_width(alignment, 64)
+        }
+        if as_untyped_ptr:
```

Comment:
I don't think that this is good practice to have `elem_type` required when requesting an untyped pointer.
Can we instead request `elem_type: Attribute | None`, so we don't have to provide it?

Reply:
The Alloca op always needs to know the type (because it needs to allocate enough memory), the `as_untyped_ptr` argument refers to whether the returned pointer carries type information or not. These are the only two valid ways to use `alloca`:

```
// type info encoded in return type
%0 = "llvm.alloca"(%len) {"alignment" = 0 : i64} : (i64) -> !llvm.ptr<i32>
// type info encoded in attributes
%0 = "llvm.alloca"(%len) {"alignment" = 0 : i64, "elem_type" = i32} : (i64) -> !llvm.ptr
```

Reply:
Oh okay, thanks for the explanation!